src/KeyringController.ts

```diff
@@ -1053,6 +1123,66 @@ class KeyringController extends EventEmitter {
 
     return keyring;
   }
+
+  /**
+   * Ensure that the vault is encrypted with the latest encryption method available.
+   *
+   * The vault is left unchanged if no password is available, if the encryptor does not support
+   * the vault update, or if the vault is already encrypted with the latest encryption method.
+   *
+   * @returns The updated encrypted vault.
+   */
+  async #updateVaultEncryption(): Promise<void> {
```

Comment:
Thoughts on moving this to the extension instead? Not sure what advantage it has at this layer. The clients are already use separate encryptors and separate encryption configuration.

Moving this to the client would also make it easier to extract encryption from the keyring controller altogether. That's something the accounts team was working on, to support encrypting more types of data, not just keyrings.

Reply:
As `eth-keyring-controller` instance is unaccessible from the extension, we could update the `keyring-controller` store (and thus also the extension storage), but as `eth-keyring-controller` does not listen to `keyring-controller` state updates (it is only the other way around) the two controllers would then be in an inconsistent state. I'm not sure this would represent a problem though - I have to look better into that, but this might need some change in `keyring-controller` in case we want to handle this in the extension

Reply:
Perhaps the extension can use the encryptor to update the vault in the storage prior to keyring controller initialisation

Reply:
Hmm, I was thinking this was an operation on the encryptor itself, but it's not. I see what you mean.

Reply:
Yes, exactly. Indeed, this is only needed to be sure that the vault gets migrated even if the user doesn't make any action on the keychain

Reply:
Gotcha, makes sense. Instead of having this custom migrate function then, could we accomplish this by calling `persistAllKeyrings()` in `unlockKeyrings`?

Reply:
We could also call `persistAllKeyrings`, but:
- it should be done after the unlock or we'll overwrite the existing vault with an empty one
- EthKeyringController would have no way of knowing if the vault is already up to date, so it would repeat the operation every time we unlock the vault, which is not great

Reply:
In https://github.com/MetaMask/KeyringController/pull/293/commits/29713e605d2009ebd3a62f66c7ee829d9db96402 we now:
- Execute the whole `unlockKeyrings`, so that the keyrings array is up to date with the saved vault
- Check the current vault against the one returned by `updateVault`
- Call `persistAllKeyrings` in case they differ

Reply:
Looks great!

Reply:
Looks great!